core/tests/protractor_utils/AdminPage.js

```diff
@@ -48,28 +48,46 @@ var AdminPage = function() {
     '.protractor-test-unfinished-one-off-jobs-rows'));
   var unfinishedOffJobIDClassName = (
     '.protractor-test-unfinished-one-off-jobs-id');
+  var explorationElements = element.all(by.css(
+    '.protractor-test-reload-exploration-row'));
+  var reloadCollectionButtons = element.all(by.css(
+    '.protractor-test-reload-collection-button'));
+  var explorationTitleLocator = by.css(
+    '.protractor-test-reload-exploration-title');
```

Comment:
We have defined selectors in multiple ways in the codebase:
1. var X = element(by.css(..))
2. someElement.element(by.css(...))
3. element.all(by.css(...))
4. element(by.css('...' + someScopeVariable + '...'))

Considering the three thoughts:
1. In our codebase the `element(..)` call contains locators in all the places.
2. One of the goals to define selectors/locators on top is to avoid duplicate constant `string` part (`.protratcor-test...`) in multiple places. For the 2nd way (above) it won't be possible to define a selector on top, so it would be better to define locators to avoid duplicate strings in multiple places.
3. To catch the cases where we can define the selectors/locators on top of would be simpler to identified using locators, instead of selectors.

```
var explorationTitle = explorationElement.element(by.css(
  '.protractor-test-reload-exploration-title');
````

This won't work as `explorationElement` is a local variable (i.e, the param of the function). Also, in general, this won't work in all such places as the parent-selector can be defined dynamically in the local scope. Some e.g:

```
function(rteElem) {
   rteElem(paragraphLocator)
}

function(someVar) {
var elem = element(by.css('...' + someVar);

await elem.element(someLocator).click();
}
```
I think we can use eslint to find such cases but it also somewhat depends on developers to define selectors/locators in the best possible ways. (As eslint is not that smart and can be easily fooled.)

Maybe we can write the error message in a better way so that it can help developers understand define locator/selector in a better way? (not sure!) Let me know what you think?

Reply:
Here is the current policy for grouping selectors at the tops of files from the [E2E wiki page](https://github.com/oppia/oppia/wiki/End-to-End-Tests#codeowner-checks):

> * [ ] All element selectors, e.g. `element(by.css('.protractor-test-my-element'))`, need to be at the top of the file. There are a few exceptions:
>
>    * Keeping selectors with the code that uses them is okay in some utility files where the utilities do not generally share selectors.
>    * When you are chaining selectors, only the root selector (first in the chain) needs to be at the top of the file.

It seems like in this PR, we are implementing a lint check that doesn't actually enforce this policy. For example, we are requiring that locators be at the tops of files, not selectors.

I understand that it's easier to write the linter if we change the policy to be about locators instead of selectors, but I think changing our policies to be about locators instead of selectors will make the code less readable. With selectors grouped at the top, we have:

```js
action.click('submit button', submitButton);
```

With locators grouped at the top, we have:

```js
action.click('submit button', element(submitSelector));
```

I think the former is much cleaner and easier to read, though please let me know if you disagree!

I'm open to changing our style policies somewhat to make writing lint checks easier, but I think the change proposed in this PR compromises readability too much. I'm open to being convinced otherwise though!

Reply:
> action.click('submit button', element(submitSelector));

This won't be possible in cases where we are using selectors inside an element i.e, `X.element(by.css(...))`.

I agree with both of you and I think it would be better to define selectors in places where it's possible and locators where selectors are not possible! This will help us define all the `string` constants at the top! 

Logic:
if `element(...)` || `element.all(...)`: define selector on top
if `X.element(by.css(...))` || `X.element(by.css(...))`: define locator on top

@U8NWXD Do you think this should be fine^?

@AdityaDubey0, is it possible to implement the above idea? Let me know if you need any help! [I can help you write the rule clearly!]

Reply:
@DubeySandeep Yes, it is possible I will try to modify my lint check according to the above idea.
@U8NWXD Can you please confirm is it fine to modify the lint check according to the above idea?

Reply:
Yes, @DubeySandeep and I agreed offline that the above approach is essentially correct. Specifically, the main goal is:

> define selectors in places where it's possible and locators where selectors are not possible

The trick is turning "where it's possible" into something we can write lint checks for. Here's what we agreed on in broad strokes:

* if `element(...)` or `element.all(...)`: define selector on top
* if `X.element(by.css(...))` or `X.element(by.css(...))`: define locator on top
* if the string inside `by.css` is not literal: allow defining locator and selector anywhere